hxselect.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXSELECT" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxselect \- extract elements or attributes that match a (CSS) selector
.SH SYNOPSIS
.B hxselect
.RB "[\| " \-i " \|]"
.RB "[\| " \-c " \|]"
.RB "[\| " \-l
.IR language " \|]"
.RB "[\| " \-s
.IR separator " \|]"
.I selectors
.SH DESCRIPTION
.B hxselect
reads a well-formed XML document and outputs all elements and attributes
that match one of the CSS selectors that are given as an argument. For example
.d
hxselect ol li:first-child
.e
selects the first li (list item in XHTML) in an ol (ordered list).
.PP
If there are multiple selectors, they must be separated by commas. For
example,
.d
hxselect p + ul, blockquote ol
.e
selects all ul elements that follow a p and all ol elements that are
descendants of a blockquote element.
.PP
The command operates on the standard input.
.PP
.B hxselect
assumes that class selectors (".foo") refer to an attribute called
"class" and that ID selectors ("#foo") refer to an attribute called
"id".
.PP
The experimental attribute node selector
.RI '::attr( name )'
is supported and selects the attribute of that name.
.PP
Comments and processing instructions are ignored, i.e., they are read
but never written.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-i
Match case-insensitively. Useful for HTML and some other
SGML-based languages.
.TP
.B \-c
Print content only. Without \fB\-c\fR, the start and end tag of the
matched element are printed as well; with \fB\-c\fR only the contents
of the matched element are printed. If an attribute rather than an
element is selected (::attr() selector), only the value of the
attribute is printed.
.TP
.BI \-l " language"
Sets the default language, in case the root element doesn't
have an xml:lang attribute (default: none). Example:
.B \-l en
.TP
.BI \-s " separator"
A string to print after each match (default: empty). Accepts
C-like escapes. Example: 
.B \-s \(aq\\\\n\\\\n\(aq
to print an empty line after each match.
.SH OPERANDS
The following operand is supported:
.TP
.I selectors
One or more comma-separated selectors. Most selectors from CSS level 3
are supported, with the exception of selectors that require
interaction (e.g., ':active') or layout (e.g., ':first-line).
.SH "SEE ALSO"
.BR asc2xml (1),
.BR xml2asc (1),
.BR hxnormalize (1),
.BR hxremove (1),
.BR UTF-8 " (RFC 2279)"


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxselect.c
/*
 * select -- extract elements matching a selector
 *
 * Assumes that class selectors (".foo") refer to an attribute called
 * "class".
 *
 * Assumes that ID selectors ("#foo") refer to an attribute called
 * "id".
 *
 * Options:
 *
 * -l language
 *
 *     Sets the default language, in case the root element doesn't
 *     have an xml: lang attribute. Example: -l en. Default: none.
 *
 * -s separator
 *
 *     A string to print after each match. Accepts C-like escapes.
 *     Example: -s '\n\n' to print an empty line after each match.
 *     Default: empty string.
 *
 * -i
 *
 *     Match case-insensitively. Useful for HTML and some other
 *     SGML-based languages.
 *
 * -c
 *
 *     Print content only. Without -c, the start and end tag of the
 *     matched element are printed as well; with -c only the contents
 *     of the matched element are printed.
 *
 * TODO: Escape double quotes when printing an attribute that matches
 * ::attr().
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * Copyright © 2001-2017 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 5 Jul 2001
 * Version: $Id: hxselect.c,v 1.10 2017/11/24 09:50:25 bbos Exp $
 *
 **/
#include "config.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <stdbool.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
# ifndef HAVE_STRSTR
#  include "strstr.e"
# endif
#endif
#include "types.e"
#include "tree.e"
#include "selector.e"
#include "heap.e"
#include "errexit.e"
#include "html.e"
#include "scan.e"
#include "selmatch.e"


static Tree tree = NULL;			/* Current elt in tree */
static Selector selector;			/* The selector to match */
static bool content_only = false;		/* Omit start/end tag */
static string separator = "";			/* Printed between matches */


/* print_starttag -- print a start tag */
static void print_starttag(Node *n)
{
  pairlist p;

  printf("<%s", n->name);
  for (p = n->attribs; p; p = p->next) printf(" %s=\"%s\"", p->name, p->value);
  putchar('>');
}


/* printsep -- print the separator string, interpret escapes */
static void printsep(const string separator)
{
  string s = separator;
  int c;

  while (*s) {
    if (*s != '\\') putchar(*(s++));
    else if ('0' <= *(++s) && *s <= '7') {
      c = *s - '0';
      if ('0' <= *(++s) && *s <= '7') {
	c = 8 * c + *s - '0';
	if ('0' <= *(++s) && *s <= '7') c = 8 * c + *s - '0';
      }
      putchar(c); s++;
    } else
      switch (*s) {
      case '\0': putchar('\\'); break;
      case 'n': putchar('\n'); s++; break;
      case 't': putchar('\t'); s++; break;
      case 'r': putchar('\r'); s++; break;
      case 'f': putchar('\f'); s++; break;
      default: putchar(*(s++)); break;
      }
  }
}


/* print_recursively -- print the element t, its children and sisters */
static void print_recursively(Tree t)
{
  if (!t) return;

  switch (t->tp) {
  case Element:
    print_starttag(t);
    print_recursively(t->children);
    printf("</%s>", t->name);
    break;
  case Text: printf("%s", t->text); break;
  case Comment: printf("<!--%s-->", t->text); break;
  case Declaration: assert(!"Cannot happen"); break;
  case Procins: printf("<?%s>", t->text); break;
  case Root: print_recursively(t->children); break;
  default: assert(!"Cannot happen");
  }

  print_recursively(t->sister);
}


/* print_tree -- print the element t, with or without its tags */
static void print_tree(Tree t)
{
  assert(t->tp == Element);
  if (!content_only) print_starttag(t);
  print_recursively(t->children);
  if (!content_only) printf("</%s>", t->name);
  printsep(separator);
}


/* match_pseudoelts -- print pseudo-elements of node t that match the selector */
static void match_pseudoelts(Tree t)
{
  pairlist p;

  assert(selector->pseudoelts);

  /* ::attr() is the only pseudo-element we can handle so far */
  if (selector->pseudoelts->type == AttrNode) {
    assert(selector->combinator == Child && selector->context);
    p = t->attribs;
    while (p && !same(p->name, selector->pseudoelts->s)) p = p->next;
    if (p && matches_sel(t, selector->context)) {
      if (!content_only) printf("%s=\"", p->name);
      printf("%s", p->value);
      if (!content_only) printf("\"");
      printsep(separator);
    }
  }
}


/* walk_tree -- find all nodes in the tree that match the selector */
static void walk_tree(Tree t)
{
  if (!t) return;

  switch (t->tp) {
  case Element:
    if (selector->pseudoelts) match_pseudoelts(t);
    else if (matches_sel(t, selector)) print_tree(t);
    walk_tree(t->children);
    break;
  case Text: break;
  case Comment: break;
  case Declaration: break;
  case Procins: break;
  case Root: walk_tree(t->children); break;
  default: assert(!"Cannot happen");
  }

  walk_tree(t->sister);
}



/*********************** Parser callback API ***********************/

/* handle_error -- called when a parse error occurred */
static void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
}


/* handle_start -- called before the first event is reported */
static void* handle_start(void)
{
  tree = create();
  return NULL;
}


/* handle_end -- called after the last event is reported */
static void handle_end(void *clientdata)
{
  walk_tree(tree);	       /* Find all matches for the selector */
  tree_delete(tree);	       /* Clean up */
}


/* handle_comment -- called after a comment is parsed */
static void handle_comment(void *clientdata, const string commenttext)
{
  free(commenttext);
}


/* handle_text -- called after a text chunk is parsed */
static void handle_text(void *clientdata, const string text)
{
  tree = tree_append_text(tree, text);
}


/* handle_decl -- called after a declaration is parsed */
static void handle_decl(void *clientdata, const string gi, const string fpi,
			const string url)
{
  free(gi);
  free(fpi);
  free(url);
}


/* handle_pi -- called after a Processing Instruction is parsed */
static void handle_pi(void *clientdata, const string pi_text)
{
  free(pi_text);
}


/* handle_starttag -- called after a start tag is parsed */
static void handle_starttag(void *clientdata, const string name,
			    pairlist attribs)
{
  tree = tree_push(tree, name, attribs);	/* Add to tree */
  free(name);
}


/* handle_emptytag -- called after an empty tag is parsed */
static void handle_emptytag(void *clientdata, const string name,
			    pairlist attribs)
{
  tree = tree_push(tree, name, attribs);	/* Add to tree */
  tree = tree_pop(tree, name);			/* Remove from tree again */
  free(name);
}


/* handle_endtag -- called after an endtag is parsed (name may be "") */
static void handle_endtag(void *clientdata, const string name)
{
  tree = tree_pop(tree, name);
  free(name);
}


/* handle_endincl -- called at the end of an included file */
/* void handle_endincl(void *clientdata) */

/********************* End of parser callbacks *********************/


/* usage -- print usage message and exit */
static void usage(const conststring progname)
{
  errexit("Usage: %s [-v] [-i] [-c] [-l language] [-s separator] selector\n",
	  progname);
}


int main(int argc, char *argv[])
{
  string s;
  int c;

  /* Command line options */
  while ((c = getopt(argc, argv, "icl:s:v")) != -1) {
    switch (c) {
    case 'c': content_only = true; break;
    case 'l': init_language(optarg); break;
    case 's': separator = optarg; break;
    case 'i': set_case_insensitive();  break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    case '?': usage(argv[0]); break;
    default: assert(!"Cannot happen");
    }
  }

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(handle_start);
  set_end_handler(handle_end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse the selector */
  if (optind >= argc) usage(argv[0]);		/* Need at least 1 arg */
  for (s = newstring(argv[optind++]); optind < argc; optind++)
    strapp(&s, " ", argv[optind], NULL);
  selector = parse_selector(s, &s);
  if (*s) errexit("Syntax error at \"%c\"\n", *s);

  /* Walk the tree */
  yyin = stdin;
  if (yyparse() != 0) exit(3);
  return 0;
}
